allensdk/brain_observatory/ecephys/_channels.py

```diff
@@ -0,0 +1,107 @@
+from typing import List, Optional
+
+import numpy as np
+import pandas as pd
+from pynwb import NWBFile
+
+from allensdk.brain_observatory.ecephys._channel import Channel
+from allensdk.brain_observatory.ecephys.utils import clobbering_merge
+from allensdk.core import DataObject, NwbReadableInterface, \
+    JsonReadableInterface
+
+
+class Channels(DataObject, NwbReadableInterface, JsonReadableInterface):
+    """A set of channels"""
+
+    def __init__(self, channels: List[Channel]):
+        super().__init__(name='channels', value=channels)
+
+    @classmethod
+    def from_json(cls, channels: dict) -> "Channels":
+        channels = [Channel(**{f'{"impedance" if k == "impedence" else k}': v
+                               for k, v in channel.items()})
+                    for channel in channels]
+        return Channels(channels=channels)
+
+    def to_dataframe(self, external_channel_columns=None,
+                     filter_by_validity=True) -> pd.DataFrame:
+        """
+
+        Parameters
+        ----------
+        external_channel_columns
+        filter_by_validity: Whether to filter channels based on whether
+            the channel is marked as "valid_data"
+
+        Returns
+        -------
+
+        """
+        channels = [channel.to_dict()['channel'] for channel in self.value]
+        channels = pd.DataFrame(channels)
+        channels = channels.set_index('id')
+        channels = channels.drop(columns=['impedance'])
+
+        if external_channel_columns is not None:
+            external_channel_columns = external_channel_columns()
+            channels = clobbering_merge(channels, external_channel_columns,
+                                        left_index=True, right_index=True)
+
+        if filter_by_validity:
+            channels = channels[channels['valid_data']]
+            channels = channels.drop(columns=['valid_data'])
+        return channels
+
+    @classmethod
+    def from_nwb(cls, nwbfile: NWBFile,
```

Comment:
I do not see the corresponding to_nwb method

Reply:
Due to the way the data needs to be written to nwb, all channels are written at once to nwb. So `to_nwb` is in `Channels` class.